Further refinements of the mempool specification (#9)
cometbft/knowledge-base: protocols/mempool-overview.md

Then, the interactions with the consensus and client application are detailed.
A formalization of the mempool follows.
We close this document with a proof that the current implementation of the mempool in CometBFT (as of v0.38.0-rc1) is correct.

## Overview

The mempool acts as an entry point to consensus.
It permits to disseminate transactions from one node to another, for their eventual inclusion in the blockchain.
To this end, the mempool maintains a replicated set, or _pool_, of transactions.
Transactions in the mempool are consumed by consensus to create the next proposed block.
The mempool is refreshed once a new block in the blockchain  is decided.

A transaction can be received from a local client, or a remote disseminating process.
Each transaction is subject to a test by the client application.
This test verifies that the transaction is _valid_.
Such a test provides some form of protection against byzantine agents, whether they be clients or other system nodes.
It also serves to optimize the overall utility of the blockchain.
Validity can be simply syntactical which is stateless, or a more complex verification that is state-dependent.

> If the transaction is valid, the local process further propagates it in the system using a gossip (or an anti-entropy) mechanism.

Reviewer: The more I think, the less it appears that we are using a gossip protocol here, ehe.

Essentially, the dissemination is doing by flooding in a partially and randomly-connected network overlay.

It provides strong global guarantees.
In particular, once a transaction is added to the mempool, it is guaranteed to be available at any later point in time.

## Interactions

In what follows, we present the interactions of the mempool with other parts of the CometBFT protocol.
Some of the specificities of the current implementation (`CListMempool`) are also detailed.
For further information about the current implementation, the reader may consult the overview [document](https://github.com/cometbft/knowledge-base/blob/main/protocols/mempool/v0/mempool-v0.md), as well as the quint [specification](https://github.com/cometbft/knowledge-base/pull/11/files).

**RPC server**
To add a new transaction to the mempool, a clients may submit it through an appropriate RPC endpoint.
This endpoint is offered by some of the system nodes (but not necessarily all of them).

**Gossip protocol** 
Transactions can also be received from other nodes, through a gossiping mechanism.

**ABCI application**
As pointed above, the mempool should only store and disseminate  valid transactions.
It is up to the ABCI (client) application to define whether a transaction is valid.

> Transactions received locally are sent to the application to be validated, through the `checkTx` method from the mempool ABCI connection.

Reviewer: ```suggestion
Transactions received locally are sent to the application to be validated, through the `CheckTx` method from the mempool ABCI connection.
```

The state of the application is updated when consensus commits a block of transactions.
When this happens, the transactions still in the mempool have to be validated again.
We further detail this mechanism below.

**Consensus**
The consensus protocol consumes transactions stored in the mempool to build blocks to be proposed.
To this end, consensus requests from the mempool a list of transactions which abide by certain limits (namely, total number of transactions included, or total size in bytes).
In the current implementation, the mempool is a list of transactions.
Such a call returns the longest prefix of the list that is matching the requirements.
Notice that at this point the transactions returned to consensus are not removed from the mempool.
This comes from the fact that the block is proposed but not decided yet.

Proposing a block is the prerogative of the nodes acting as validators.
At all the nodes (validators or not), consensus is also responsible for committing blocks of transactions to the blockchain.
Once a block is committed, all the transactions included in the block are removed from the mempool.

> This happens with an `update` call to the mempool.

Reviewer: Are you using not-capitalized naming here (update, checkTx etc.)?

I wonder whether using the same names as in the code and the rest of the documentation it is better. Namely, `CheckTx` is a public method and the name of the ABCI method.

received from a peer in the mempool protocol.
Transactions that are validated by the application are appended to the local
copy of the mempool.

> Transactions considered invalid by the application are drooped, therefore are

Reviewer 2: ```suggestion
Transactions considered invalid by the application are dropped, therefore are
```

The height of the ledger at $p$ is the index of the first null entry; denoted $p.height$.
Operation $submit(txs, i)$ attempts to write the set of transactions $txs$ to the $i$-th entry of the ledger.
The (history) variable $p.submitted[i]$ holds all the transactions (if any) submitted by $p$ at height $i$.
By extension, $p.submitted$ are all the transaction submitted by $p$.
A transaction is committed when it appears in one of the entries of the ledger.
We write $p.committed$ the committed transactions at $p$.

As standard, the ledger ensures that:  
* _(Gap-freedom)_ There is no gap between two entries at a correct process:  
$\forall i \in 	\mathbb{N}. \forall p \in Correct. \square(p.ledger[i] \neq \bot \implies (i=0 \vee p.ledger[i-1] \neq \bot))$;  
* _(Agreement)_ No two correct processes have different ledger entries; formally:  
$\forall i \in 	\mathbb{N}. \forall p,q \in Correct. \square((p.ledger[i] = \bot) \vee (q.ledger[i] = \bot) \vee (p.ledger[i] = q.ledger[i]))$;  
* _(Validity)_ If some transaction appears at an index $i$ at a correct process, then a process submitted it at that index:  

> $\forall p \in Correct. \exists q \in Processes. \forall i \in 	\mathbb{N}. \square(tx \in p.ledger[i] \implies tx \in \bigcup_q q.submitted[i]$).

Reviewer: I think we don't need the \bigcup here.

Reviewer: Actually, following the english definition of this property, for every p and i we need a q for which tx \in q.submitted[i]. Namely, the \exist q should be moved to the right part and the union doesn't make sense here.

> * _(Termination)_ If a correct process submits a block at its current height, eventually its height get incremented:  

Reviewer: The definition does not really uses the definition of block up to here. Processes submit transactions.

Reviewer: What it looks to be the spirit of this property is that the height of the blockchain increases, meaning that new blocks of transactions are appended to it, if any correct process submits new transactions to append to the blockchain.

> At a process $p$, we shall write $p.var$ the local variable $var$ at $p$.

Reviewer 2: This is just personal preference: I would change the word "process" to "node". I think in this context it makes more sense. A node has a ledger which is at some height. A node sends and receives messages.

Author: I prefer to stick with the term _process_, as this is the one commonly encountered in specifications (see e.g., quint ones).

**Ledger.**
We use the standard definition of (BFT) SMR in the context of blockchain, where each process $p$ has a ledger, written $p.ledger$.
At process $p$, the $i$-th entry of the ledger is denoted $p.ledger[i]$.
This entry contains either a null value ($\bot$), or a set of transactions, aka., a block.
The height of the ledger at $p$ is the index of the first null entry; denoted $p.height$.
Operation $submit(txs, i)$ attempts to write the set of transactions $txs$ to the $i$-th entry of the ledger.

> The (history) variable $p.submitted[i]$ holds all the transactions (if any) submitted by $p$.

Reviewer 3: ```suggestion
The (history) variable $p.submitted[i]$ holds all the transactions (if any) submitted by $p$ at height $i$.
```

By extension, $p.submitted$ are all the transaction submitted by $p$.

> A transaction is committed when it appears in some entries of the ledger.

Reviewer 2: ```suggestion
A transaction is committed when it appears in at least one entry of the ledger.
```

We write $p.committed$ the committed transactions at $p$.

> As standard, the ledger ensures that:  

Reviewer 2: To be more precise, we could say that all these properties are ensured by the consensus module, not by the ledger.

Author: I believe it is better to distinguish (one-shot) consensus from the ledger (SMR). Even if, it is common that the two notions are used interchangeably. This is clear from the quint specification.

* _(Gap-freedom)_ There is no gap between two entries at a correct process:  
$\forall i. \forall p \in Correct. \square(p.ledger[i] \neq \bot \implies (i=0 \vee p.ledger[i-1] \neq \bot))$;  
* _(Agreement)_ No two correct processes have different ledger entries; formally:  
$\forall i. \forall p,q \in Correct. \square((p.ledger[i] = \bot) \vee (q.ledger[i] = \bot) \vee (p.ledger[i] = q.ledger[i]))$;  
* _(Validity)_ If some transaction appears at an index $i$ at a correct process, then a process submitted it at that index:  
$\forall p \in Correct. \forall i. \square(tx \in p.ledger[i] \implies tx \in \bigcup_q q.submitted[i]$).
* _(Termination)_ If a correct process submits a block at its current height, eventually its height get incremented:  
$\forall p \in Correct. \square((h=p.height \wedge p.submitted[h] \neq \varnothing) \implies \lozenge(p.height>h))$  

**Mempool.**
A mempool is a replicated set of transactions.
At a process $p$, we write it $p.mempool$.
We also define $p.hmempool$, the (history) variable that tracks all the transactions ever added to the mempool by process $p$.
Below, we list the invariants of the mempool (at a correct process).

> The mempool is used as an input for the ledger:  

Reviewer 3: We need to be careful here. `PrepareProposal`, in its current form (as of `v0.37.x`) violates this property, as the Application can include TXs directly in the proposal (and those TXs aren't added to the mempool)

Reviewer: Can we find a way to have this case represented here? Even if we don't consider `PrepareProposal`, the ledger (blockchain) contains information, that can be called a transaction or not, which was never been part of the mempool. For instance, it contains evidence of misbehavior, handled by the evidence protocol; it contains validator changes, consensus parameters changes, etc.

So, in a general way, we should allow for the ledger and blocks to contain transactions that were not retrieved from the mempool, even though we don't want to model more complex possibilities introduced with ABCI++.

In blockchain, a tx is (or not) valid in a given state.
That is a transaction can be valid (or not) at a given height of the ledger.
To model this, consider that $p.ledger.valid(tx)$ is such a check for the current height of the ledger at process $p$ (ABCI call).
Our third invariant is that only valid transactions are present in the mempool:  

> **INV3.** $\forall tx, \forall p \in Correct. \square(tx \in p.mempool \implies p.ledger.valid(tx))$

Reviewer 3: Here, also, we need to add a footnote or something similar clarifying that operators may disable `ReCheckTx` via config change

Author: This paramter is now mentioned later in the text.

Finally, we require some progress from the mempool.
Namely, if a transaction appears at a correct process then eventually it is committed or forever invalid.  
**INV4** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies \lozenge\square(tx \in p.committed \vee \neg p.ledger.valid(tx)))$

The above invariant ensures that if a transaction enters the mempool, then it eventually leaves it at all the correct processes.
It requires from the application that the validity of a transaction converges toward some value.
That is, in every run, there is a height after which $valid(tx)$ always returns the same value.

> This happens in [Ethereum](https://github.com/ethereum/go-ethereum/blob/5c51ef8527c47268628fe9be61522816a7f1b395/light/txpool.go#L401) as a transaction is always valid until a transaction from the same address executes with the same or higher nonce.

Reviewer 3: You might also want to link the new requirement in ABCI spec, once it lands

A simpler way to satisfy this requirement is by having $valid(tx)$ deterministic and stateless (e.g., a syntactic check).

**Practical considerations.**
In practice, as it requires to traverse the whole ledger, INV2 is too expensive.

> Instead, we would like to maintain this only over the last $\alpha$ committed transactions, for some parameter $\alpha$.

Author: I think the wording is fine.

Given a process $p$, we write $p.lcommitted$ the last $\alpha$ committed transactions at $p$.
Invariant INV2 is replaced with:  

> **INV2a.** $\forall tx. \forall p \in Correct. \square(tx \in p.lcommitted \implies tx \notin p.mempool)$

Reviewer: I think this case is tracked by the $p.hmempool$.

> **INV2a.** $\forall tx. \forall p \in Correct. \square(tx \in p.lcommitted \implies tx \notin p.mempool)$

Reviewer: I think this case is tracked by the $p.hmempool$.

> **INV2a.** $\forall tx. \forall p \in Correct. \square(tx \in p.lcommitted \implies tx \notin p.mempool)$

Author: Indeed.

Another practical concern is with INV3.

> This invariant requires to have a green light from the client application before adding a transaction to the mempool.

Reviewer 4: ```suggestion
This invariant requires to have a green light from the client application before adding/maintaining a transaction to/in the mempool.
```

> This invariant requires to have a green light from the client application before adding a transaction to the mempool.

Reviewer 4: adding/maintaining?

For efficiency, such a validation needs to be made at most $\beta$ times per transaction at each height, for some parameter $\beta$.

> Ideally, $\beta$ equals $1$.

Reviewer 3: If I understood this part correctly, $\beta$ cannot be 1, as the first time a TX is received at a process $p$, CometBFT calls `CheckTx` upon reception, and then `ReCheckTx` upon decision (if the Tx is not in the decided block).
Unless you consider that the call to `ReCheckTx` actually belongs to $h + 1$...

Reviewer 4: I don't think he is counting ReCheckTx here at all.

Author: A recheck happens at the next height. The model does not distinguish between checks and rechecks. Both calls increment `p.valid[tx]`.

In practice, $\beta = f(T)$ for some function f of the maximal number of transactions T submitted between two heights.

> Given some transaction $tx$, variable $p.valid[tx]$ tracks the number of times the application was asked at the current height.

Reviewer: I think $p.valid$ here is not a great name. I would use something less "definitive", like $p.checks$ or something like that.

Reviewer: All in all, this is almost an implementation detail. Not sure whether we need this modeled here.

Invariant INV3 is replaced with:  
**INV3a.** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies p.valid[tx] \in [1, \beta])$

> ## Implementation in CometBFT (as of v0.38.0-alpha.2)

Reviewer 3: ```suggestion
## Implementation in CometBFT (as of v0.38.0-rc1)
```

Author: Fixed.

The mempool is implemented in `clist_mempool.go`, in the `CListMempool` data type.
`CListMempool` uses a single variable for the two mechanisms covered in the previous section.
Below, we present this approach in detail then establish its correctness.

**Algorithm**
For starters, we explain at a high-level the logic in `CListMempool`.
Variables `txs` and `cache` respectively hold the mempool and the validity cache in a FIFO and LRU list.
We omit the use of `txsMap`, assuming that a transaction is in `txs` iff it is also in `txsMap` (see [this](https://github.com/cometbft/cometbft/pull/890) fix).
For simplicity, we consider that 

> _(i)_ the mempool is never full, 

Author: This is a high-level proof. If we have to delve into all the subtelities (and interleavings) of `clist_mempool.go`, this is just an untractable problem.

Hence, we make some simplifications, e.g., txsMap is not there (because anyway, we must now have $tx \in txs iff tx \in txsMap$.

> _(ii)_ invalid transactions are _not_ kept in the cache, that is the parameter `KeepInvalidTxsInCache` is always set to false, and

Author: Fixed.

_(iii)_ the `flush` operation is never called.

> In addition, we shall assume that a finite amount of transactions is received by the system, and that if a transaction is received at a correct process, then it is eventually received at all the correct processes (thanks to the gossip layer in `mempool/reactor.go`).

Reviewer 3: For clarity, consider putting all 5 assumptions in bullet points

Then, let us consider some transaction $tx$.
According to the logic in `clist_mempool.go`,

$tx$ is added to the cache at time $t$ (in short, @t) if  
(1) $tx$ is received and not already there @t (`checkTx`, [l237](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L237)), or  

> (2) $tx$ is committed and valid @$t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L598)).

Reviewer 3: ```suggestion
(2) $tx$ is committed and valid @t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L598)).
```

Reviewer 3: "and valid". I remember having a discussion with @hvanz on this. My impression at the time was that we probably don't want that.

Do you have a clearer idea now on whether we want to remove failed, committed TXs from the cache?

$tx$ is removed from the cache @t if  

> (3) $tx$ was received @t'<t, re-checked and invalid t'<@t''<t (`resCbRecheck`, [l477](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L4747)), or  

Author: The removal occurs at time t.

(4) $tx$ is ejected from the cache as it is the last transaction wrt. cache and a new transaction is added, or  

> (5) $tx$ is committed and invalid @t'<t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L601)).

Reviewer 3: "and invalid". I remember having a discussion with @hvanz on this. My impression at the time was that we probably don't want that.

Do you have a clearer idea now on whether we want to remove failed, committed TXs from the cache?

$tx$ is removed from the mempool @t if  
(7) $tx$ is committed @t (`RemoveTx` after `removeTxByKey` after `update`, [l614](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L614)), or  
(8) $tx$ is re-checked and invalid @t'<t (`removeTx` after `resCbRecheck`, [l474](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L474)).

In addition, following the logic of CometBFT,  
- check and re-check operations are asynchronous,
- before committing transactions, checks and re-checks are frozen and all the pending calls return, and
- upon commit, all the transactions remaining in the mempool are re-checked.

**Correctness**
We now establish that the above algorithm implements the mempool abstraction.
For invariants INV2a and INV3a, we show that **there exist** such an $\alpha$ and $\beta$, but we do not look at characterizing them precisely.
Regarding INV4, we shall assume that step (1) occurs infinitely often and not once upon the reception of the transaction.

> This models that a CometBFT client may re-submit a failed transaction forever until it succeeds (if it does).

Reviewer: What is a "failed" transaction here? I didn't fully understand the statement before, referring to (1).

Transaction $tx$ is added to $p.lcommitted$ when step (2).
It is removed when (4) takes place.
Step (6) adds the transaction to the mempool.
The transaction is removed from the mempool upon (7) and (8).

Consider some correct process $p$.
Below, we show in order that all the invariants of the mempool are maintained at $p$:  

- INV1.
Clear from the code base, as the consensus input corresponds to a call to `Execution:CreateProposalBlock` which itself is calling `ClistMempool:ReapMaxBytesMaxGas`, returning a subset of $p.mempool$.

- INV2a.
If $tx$ is in $p.lcommitted$, it was added due to step (2).
This step occurs only if the transaction was committed at some height.
Furthermore, when this happens step (7) is taken.
It follows, that the transaction is removed from the mempool, as required.

- INV3a.
Consider some height $h$ for which $tx$ is in the mempool and $p.valid[tx]>0$.
For any transaction $tx$, this counter is set to $0$ initially.
Hence, $p.valid[tx]$ was incremented at some height $h' \leq h$.
Consider the point in time $t'$, where p moves to height $h$.
If $p.valid[tx]$ equals 0 at that time, then it was incremented at height h.

> From (6), it is necessarily valid at height h, as requied.

Reviewer: ```suggestion
From (6), it is necessarily valid at height h, as required.
```

Otherwise, the transaction is rechecked at the current (new) height h with step (8).
Because $p.valid[tx]$ is not reset, it follows that the application considers it valid at h 

- INV4.
Consider that a transaction $tx$ enters the mempool at $p$.
If $tx$ leaves the mempool, step (7) or (8) happens.
In the former case, $tx$ is committed as required.
In the latter, $tx$ is re-checked and invalid at some height $h$.
Re-checking a transaction happens only once per height, say at time $t$ for transaction $tx$ and height $h$.
When this happens at time $t$ (line 474), the transaction is also removed from the cache (line 477) due to step (3).
By assumption, step (1) is eventually re-executed for transaction $tx$ after time $t$.
Because $tx$ is no more in the cache, it is added again after time $t$.

> Now, assume that transaction $tx$ is not invalid infinitely often at process $p$.

Reviewer 3: With the new version of the requirement to the app, we would also need to explain the "opposite" assumption here: that a tx eventually converges to be (and stay) invalid

Author: In the proof, I had omitted that eventual non-oscillation was necessary to prove INV4. This assumption is now mentionned.

> Hencce after some time, it is always valid at $p$.

Reviewer 3: ```suggestion
Hence after some time, it is always valid at $p$.
```

Observe that this is also eventually the case at all the correct processes in the system.
Name $t'$ the moment in time when this occurs.
By assumption, there are only a bounded amount of transactions submitted in the system.
After time $max(t,t')$, $tx$ is eventually submitted at some height at which it is valid and committed.

*A remark.* 
The reader might observe that in the above reasoning we do not use step (5).
This comes from the fact that stricly speaking this step is not necessary for the mempool invariants to hold.
One could consider a stronger variation of INV4 in which a transaction is eventually either 
forever invalid,
or at some height both valid and committed.

> This variation would necessitate step (5) to oust a committed yet invalid transaction from the cache.

Reviewer 3: Probably one if the AIs coming from this spec is what to do with step (5). Potentially a subject for a comm call, to see what users think, and whether removing it would break existing Apps (I don't think so, but that's jut my view).

Author: Indeed, this is an interesting discussion. 

Step (5) comes from the fact that CometBFT allows an application to report as invalid some of the committed transactions. To some extend, this is not usual. From what I see, this is not the case in [Ethereum](https://docs.alchemy.com/docs/how-to-cancel-a-private-transaction-on-ethereum), nor in the [RedBelly](https://csrg.redbellyblockchain.io/doc/ConsensusRedBellyBlockchain.pdf) blockchain, nor in [this](https://arxiv.org/pdf/1805.08429.pdf) attempt to formalize Tendermint.

Nonetheless, as we discussed, allowing this permits to hide latency as the mempool may contain transactions that depend one from another (aka., pipelining). Due to byzantine nodes, such a dependency might be broken at commit time.

Reviewer 3: From a user's perspective, (a) I would imagine myself submitting tx1, and then waiting for it to appear in a block. If the tx1 appears in a block as failed, I (as a user) would be expected to re-issue the tx again (re-try).

The alternative would be (b) for me (the user) to think: as tx1 failed in block B, I don't need to do anything: it will re-try by itself.

I find (a) much simpler.

Reviewer: I didn't get the full point here. The transaction is committed but invalid, so it is removed from the cache, ok. But it is also removed from the mempool, right? If so, how can it be included in a future block?
